server/src/main/java/org/elasticsearch/index/mapper/SyntheticSourceDocumentParserListener.java

```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.mapper;
+
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+class SyntheticSourceDocumentParserListener implements DocumentParserListener {
+    private final CustomSyntheticSourceFieldLookup customSyntheticSourceFieldLookup;
+    private final IndexSettings indexSettings;
+    private final XContentType xContentType;
+
+    private final Map<LuceneDocument, Map<String, List<StoredValue>>> ignoredSourceValues;
+
+    private State state;
+
+    SyntheticSourceDocumentParserListener(MappingLookup mappingLookup, IndexSettings indexSettings, XContentType xContentType) {
+        this.customSyntheticSourceFieldLookup = mappingLookup.getCustomSyntheticSourceFieldLookup();
+        this.indexSettings = indexSettings;
+        this.xContentType = xContentType;
+
+        this.ignoredSourceValues = new HashMap<>();
+        this.state = new Tracking();
+    }
+
+    @Override
+    public boolean isActive() {
+        return state instanceof Storing;
+    }
+
+    @Override
+    public void consume(Token token) throws IOException {
+        if (token == null) {
+            return;
+        }
+
+        this.state = state.consume(token);
+    }
+
+    @Override
+    public void consume(Event event) throws IOException {
+        if (event == null) {
+            return;
+        }
+
+        this.state = state.consume(event);
+    }
+
+    private boolean shouldSkipEvent(Event event) {
+        return switch (event) {
+            case Event.LeafValue leafValue -> {
+                var reason = customSyntheticSourceFieldLookup.getFieldsWithCustomSyntheticSourceHandling()
+                    .get(leafValue.fieldMapper().fullPath());
+
```

Comment:
Simpler: `return reason == null || (reason == CustomSyntheticSourceFieldLookup.Reason.SOURCE_KEEP_ARRAYS && leafValue.insideObjectArray() == false)`

Reply:
I have a preference for my structure.